Fix JSON serialization of some PoS taggers

alexandredias3d opened the pull request:
I tried to find an alternative for serializing taggers other than using Pickle. I believe this is important to provide trained models in a safe fashion using the library.

This pull request fixes the error when using ```JSONTaggedEncoder``` on a ```NgramTagger``` with ```n``` greater than 1. Without the fix, it tries to serialize tuples and raises ```TypeError```. The workaround uses ```ast.literal_eval``` to read tuples in `repr` format. Also, this pull request can encode and decode PerceptronTagger and BrillTagger correctly.

alexandredias3d committed:
Fix typo in RegexpTagger's encode_json_obj

alexandredias3d committed:
Fix JSON encoding and decoding of NgramTaggers

Remove encode_json_obj() and decode_json_obj() from subclasses of NgramTagger. Provide serialization support to ngrams of size greater than 1: every key in _context_to_tag is converted to its repr string and loaded using ast.literal_eval. If _context_to_tag is empty, it means that all contexts were excluded since the backoff tagger covers them and the backoff tagger can be used instead.

alexandredias3d committed:
Add JSON serialization to PerceptronTagger

alexandredias3d committed:
Fix JSON serialization of Rule

Decode Rule conditions as tuples such that comparing the original and decoded version Brill Tagger'rules return True.

alexandredias3d committed:
Remove Python 2 class notation

alexandredias3d commented:
I would like to provide tests for these features. Should I write them as unit or doc tests?

Also, shouldn't the module ```tqdm``` be included in the requirements?

alvations commented:
@alexandredias3d Sorry about the `tqdm`, this shouldn't affect anything other than parallizable functions, localizing the imports inside the function would make the tqdm requirement optional only to the CLI functions. #2370

alvations commented on nltk/tag/perceptron.py:
```diff
@@ -33,9 +35,11 @@ class AveragedPerceptron(object):
+    def __init__(self, weights=None):
         # Each feature gets its own weight vector, so weights is a dict-of-dicts
-        self.weights = {}
+        self.weights = weights if weights else {}
```
👍 Much cleaner with optional weights.

alexandredias3d committed:
Add unit tests for JSON serialization of taggers

Create test_json_serializaiton to test JSON encoding and decoding of the
following taggers: Default, Affix, Regexp, Unigram, Bigram, Trigram,
Ngram, Perceptron, and Brill. Rename _regexs in RegexpTagger to _regexps
to keep consistency with the parameter name.

alexandredias3d committed:
Add myself to AUTHORS

Also, remove an item from Code Guidelines in CONTRIBUTING.md about
Python 2 support.

alexandredias3d commented:
I have one last question before I close this PR: is there a reason why the ```super``` keyword is not being used in ```sequential.py```? I have seen several cases where ```SuperClass.__init__()``` is used instead of ```super().__init__()```. I am not sure whether this is related to Python 2 support or not. If it is, I can change it to the ```super().__init__()``` notation.

alvations commented:
Yes, it's right. The `SuperClass.__init__()` patterns comes from Python 2 code. It should be safe to change it to `super().__init__()` as long as the regression tests checks out =)

BTW, awesome tests for the serialization and loading!

alexandredias3d committed:
Drop Python 2 support

Remove Python 2 class notation and change format strings to new style.

alexandredias3d commented:
@alvations I am not used to contributing to open source projects, so I have no idea if I made the right thing closing the Pull Request :grin:. Anyway, I am mentioning you so that this PR will not be forgotten.

alvations commented:
@alexandredias3d Thank you again for the contribution!! It's a really well written piece of code. 

Nope, the right thing to do is to merge the pull-request (PR) not close it =)

No worries, I've reopened it. And this PR looks good to merge (LGTM). We'll wait for @stevenbird to look through a last time and them merge it so that everyone using NLTK gets the new code =)

stevenbird commented:
Nice work @alexandredias3d. Thanks @alvations
